compiler/rustc_lexer/src/lib.rs

```diff
@@ -365,6 +370,49 @@ impl Cursor<'_> {
                 _ => self.ident_or_unknown_prefix(),
             },
 
+            // Guarded string literal (reserved syntax).
+            '#' if matches!(self.first(), '"' | '#') => {
```

Comment:
I feel like most of this match arm should be moved into a separate function, as is done for cases like `c_or_byte_string`.